clang/lib/Analysis/LifetimePsetBuilder.cpp

```diff
@@ -735,19 +734,18 @@ class PSetsBuilder : public ConstStmtVisitor<PSetsBuilder, void> {
             return;
           PSet ArgPS = getPSet(Arg);
           for (Variable V : ArgPS.vars())
-            invalidateVar(V, 1,
-                          InvalidationReason::Modified(Arg->getSourceRange()));
+            invalidateOwner(
+                V, InvalidationReason::Modified(Arg->getSourceRange()));
         },
         [&](Variable, const RecordDecl *RD, const Expr *ObjExpr) {
           const auto *RT = RD->getTypeForDecl();
           if (classifyTypeCategory(RT) != TypeCategory::Owner ||
               isLifetimeConst(Callee, QualType(RT, 0), -1))
             return;
-          PSet ArgPS = derefPSet(getPSet(ObjExpr));
-          for (Variable V : ArgPS.vars())
-            invalidateVar(
-                V, V.getOrder(),
-                InvalidationReason::Modified(ObjExpr->getSourceRange()));
+          PSet ArgPs = getPSet(ObjExpr);
```

Comment:
I'd like to see a test where an owner is invalidated by via this path (using `this`). I'm not sure that this is actually possible, because I think we don't check member functions of Owners at all.

Reply:
Hm. I haven't been able to construct a case for this, but maybe I misunderstood your intention. 
The idea was to come up with a case where an owner is invalidated by calling a non const member function on an object of the owner type ?

Reply:
I think this path is dead. We don't check member functions of Owner types (see https://github.com/mgehre/llvm-project/blob/24d454402b3f97907e1ec4e755373b5ad35e825d/clang/lib/Analysis/Lifetime.cpp#L325), and thus `this` is never an Owner. we can probably replace this by
```
[&](Variable, const RecordDecl *RD, const Expr *ObjExpr) {
  return; // 'this' is not an Owner because we don't check member functions of Owners.
}
```

Reply:
Feel free to merge after this

Reply:
Hm I think this can't be done. It is true that the body of member functions of Owners is not checked, but the code in question to be dead is still required to invalidate psets that contain the owner.
For example in the code we added as an additional test case:
 std::vector<std::vector<int>> v1;
  std::vector<int> &v2 = *v1.begin();
  auto *p1 = &v1;
  auto *p2 = &v2;
  auto i1 = v1.begin();
  auto i2 = v2.begin();
  v2.push_back(1);
  __lifetime_pset(v2); // expected-warning {{pset(v2) = ((invalid))}}
  __lifetime_pset(p1); // expected-warning {{pset(p1) = (v1)}}
  __lifetime_pset(p2); // expected-warning {{pset(p2) = (*v1)}}
  __lifetime_pset(i1); // expected-warning {{pset(i1) = (*v1)}}
  __lifetime_pset(i2); // expected-warning {{pset(i2) = ((invalid))}}
  i2 = v2.begin();
  v1.push_back({});    // expected-note {{modified here}}
  __lifetime_pset(v2); // expected-warning {{pset(v2) = ((unknown))}}

the invalidate in the PsetBuilder is required to invalidate the psets of v2, p1, p2, i1 and i2 after the push back on v1.
The pset of v1 remains (*v1)

Reply:
Ups, you are right. I now realize that the second lambda handles the "object" parameter of member function calls, which does not need to be `this`. So nothing left to do.